### PR TITLE
Block Editor: Try using DependencyExtractionWebpackPlugin

### PR DESCRIPTION
**Pull request by gziolo**
Work in progress.

I'll move to Trac as soon as Travis turns green :)

### CODE REVIEW

**Comment by sirreal** [comment]
You'll want to delete these lines:

https://github.com/WordPress/wordpress-develop/blob/f8688bb43b8e9a9478de356c086dfdab11bd1f77/tools/webpack/packages.js#L110-L124

**Comment by gziolo** [comment]
@sirreal, thanks for pointers. It solved the issue 🎉 

I had to include `@wordpress/block-directory` as a dependency. We will have to fix it on Gutenberg side first as it's hidden behind the flag and it shouldn't be necessary to be registered and enqueued in WordPress 5.3. /cc @youknowriad 

@omarreiss, we add a number of PHP files for each JS script. Is it fine to keep them next to distribution version of JS files as presented on the screenshot?

Screen Shot 2019-09-24 at 17 27 29

The same structure is present in both `build` and `src` folders.

**Comment by gziolo** [comment]
The discussion should continue on WordPress Trac: https://core.trac.wordpress.org/ticket/48154.

**Comment by gziolo** [comment]
I updated the patch with the latest changes from the `master` branch. 

In addition, I switched the output format for asset files to JSON.

Screen Shot 2019-12-04 at 11 23 24

It also looks like JSON is now required to run WordPress so we can take advantage of it:
https://make.wordpress.org/core/2019/10/15/php-native-json-extension-now-required/

File: Gruntfile.js [hunk 1]
```diff
@@ -342,6 +348,12 @@ module.exports = function(grunt) {
 					}
 				]
 			},
+			assets: {
+				expand: true,
+				cwd: WORKING_DIR + 'wp-includes/js/',
+    			src: 'dist/*.asset.php',
```
Comment (gziolo) [inline Gruntfile.js]: hmmm, I need to fix it...

File: Gruntfile.js [hunk 2]
```diff
@@ -108,8 +108,12 @@ module.exports = function(grunt) {
 			],
 			js: [
 				WORKING_DIR + 'wp-admin/js/',
+				WORKING_DIR + 'wp-includes/assets/',
 				WORKING_DIR + 'wp-includes/js/'
 			],
+			assets: [
+				WORKING_DIR + 'wp-includes/js/**/*.asset.php'
```
Comment (ntwb) [inline Gruntfile.js]: Based [on the core trac image](https://core.trac.wordpress.org/attachment/ticket/48154/Screen%20Shot%202019-12-04%20at%2016.10.08.png) the `php` assets are now stored in the `/src/wp-includes/assets/dist` folder

Comment (gziolo) [reply Gruntfile.js]: It's more complex than that.

This is where webpack generates those PHP assets, we move them to `/src/wp-includes/assets/dist` or `/build/wp-includes/assets/dist` depending where webpack operates. It's even more hard to follow when you take into account that `npm run build:dev`  operates in `src` folder, but `npm run build` uses `build` folder ...

Comment (gziolo) [reply Gruntfile.js]: This name should probably be updated to `webpack-generated-assets` :)

Comment (gziolo) [reply Gruntfile.js]: @ntwb I included some tweaks to make it easier to follow. I hope it helps:

https://github.com/WordPress/wordpress-develop/pull/102/commits/ba91cc12103ffc7d0720c0c9be0aa850d47ce5cd

**Comment by gziolo** [comment]
Resolved with https://core.trac.wordpress.org/changeset/47035.
